src/fast-instantiator.ts

```diff
@@ -0,0 +1,205 @@
+import type { IAnyClassModelType, IAnyType, IClassModelType, Instance, InstantiateContext, SnapshotIn, ValidOptionalValue } from "./types";
+import { $identifier } from "./symbols";
+import { MapType, QuickMap } from "./map";
+
+export const $fastInstantiator = Symbol.for("mqt:class-model-instantiator");
+
+export type CompiledInstantiator<T extends IAnyClassModelType = IAnyClassModelType> = (
+  instance: Instance<T>,
+  snapshot: SnapshotIn<T>,
+  context: InstantiateContext
+) => void;
+
+/**
+ * Compiles a fast function for taking snapshots and turning them into instances of a class model.
+ **/
+export const buildFastInstantiator = <T extends IClassModelType<Record<string, IAnyType>, any, any>>(model: T): CompiledInstantiator<T> => {
+  return new InstantiatorBuilder(model).build();
+};
+
+type DirectlyAssignableType = SimpleType<any> | IntegerType | LiteralType<any> | DateType;
+const isDirectlyAssignableType = (type: IAnyType): type is DirectlyAssignableType =>
+  type instanceof SimpleType || type instanceof IntegerType || type instanceof LiteralType || type instanceof DateType;
+
+class InstantiatorBuilder<T extends IClassModelType<Record<string, IAnyType>, any, any>> {
+  aliases = new Map<string, string>();
+
+  constructor(readonly model: T) {}
+
+  build(): CompiledInstantiator<T> {
+    const segments: string[] = [];
+
+    for (const [key, type] of Object.entries(this.model.properties)) {
+      if (isDirectlyAssignableType(type)) {
+        segments.push(`
+        // simple type for ${key}
+        instance["${key}"] = ${this.expressionForDirectlyAssignableType(key, type)};
+      `);
+      } else if (type instanceof OptionalType) {
+        segments.push(this.assignmentExpressionForOptionalType(key, type));
+      } else if (type instanceof ReferenceType || type instanceof SafeReferenceType) {
+        segments.push(this.assignmentExpressionForReferenceType(key, type));
+      } else if (type instanceof MapType) {
+        segments.push(this.assignmentExpressionForMapType(key, type));
+      } else {
+        segments.push(`
+          // instantiate fallback for ${key} of type ${type.name}
+          instance["${key}"] = ${this.alias(`model.properties["${key}"]`)}.instantiate(
+            snapshot?.["${key}"], 
+            context, 
+            instance
+          );
+        `);
+      }
+    }
+
+    for (const [key, _metadata] of Object.entries(this.model.volatiles)) {
+      segments.push(`
+      instance["${key}"] = ${this.alias(`model.volatiles["${key}"]`)}.initializer(instance);
+    `);
+    }
+
+    const identifierProp = this.model.mstType.identifierAttribute;
+    if (identifierProp) {
+      segments.push(`
+      const id = instance["${identifierProp}"];
+      instance[$identifier] = id;
+      context.referenceCache.set(id, instance); 
+    `);
+    }
+
+    const innerFunc = `
+      return function Instantiate${this.model.name}(instance, snapshot, context) {
```

Comment:
We'll get a nice function name in our stack traces, eh? 🚀 

Should we `camelize(name)` or something though? I can't remember if MST had any strict requirements on the name for a `types.model`. Or will this be the name of a class, which already adheres to the JS function/var/etc naming constraints?

Reply:
Right now we're only using this for class models which ya will have a safe JS function name so I think this is ok. We also don't actually have a camelize utility handy, just the one memoize function from lodash atm.